<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# Exploring and pre-processing a dataset using Pandas 


Estimated time needed: **30** minutes
    

## Objectives

After completing this lab you will be able to:

* Explore the dataset
* Pre-process dataset as required (may be for visualization)


## Introduction

The aim of this lab is to provide you a refresher on the **Pandas** library, so that you can pre-process and anlyse the datasets before applying data visualization techniques on it. This lab will work as acrash course on *pandas*. if you are interested in learning more about the *pandas* library, detailed description and explanation of how to use it and how to clean, munge, and process data stored in a *pandas* dataframe are provided in our course [**Data Analysis with Python**](https://www.coursera.org/learn/data-analysis-with-python?specialization=ibm-data-analyst) and [**Python for Applied Data Science**](https://www.coursera.org/learn/python-for-applied-data-science-ai?specialization=ibm-data-analyst)

------------


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

1. [Exploring Datasets with *pandas*](#0)<br>
2. [The Dataset: Immigration to Canada from 1980 to 2013](#1)<br>
3. [*pandas* Basics](#2) <br>
4. [*pandas* Intermediate: Indexing and Selection](#3) <br>
5. [*pandas* Filtering based on a criteria](#4)<br>
6. [*pandas* Sorting Values](#5)

</div>


# Exploring Datasets with *pandas* <a id="0"></a>

*pandas* is an essential data analysis toolkit for Python. From their [website](http://pandas.pydata.org/):
>*pandas* is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, **real world** data analysis in Python.

The course heavily relies on *pandas* for data wrangling, analysis, and visualization. We encourage you to spend some time and familiarize yourself with the *pandas* API Reference: http://pandas.pydata.org/pandas-docs/stable/api.html.


## The Dataset: Immigration to Canada from 1980 to 2013 <a id="1"></a>


Dataset Source: [International migration flows to and from selected countries - The 2015 revision](https://www.un.org/development/desa/pd/data/international-migration-flows).

The dataset contains annual data on the flows of international immigrants as recorded by the countries of destination. The data presents both inflows and outflows according to the place of birth, citizenship or place of previous / next residence both for foreigners and nationals. The current version presents data pertaining to 45 countries.

In this lab, we will focus on the Canadian immigration data.

![Data Preview](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/Module%201/images/DataSnapshot.png)

 The Canada Immigration dataset can be fetched from <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx">here</a>.

---


## *pandas* Basics<a id="2"></a>


The first thing we'll do is install **openpyxl** (formerly **xlrd**), a module that *pandas* requires to read Excel files.


In [1]:
!mamba install openpyxl==3.0.9 -y


                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['openpyxl==3.0.9']

[+] 0.0s
[+] 0.1s
pkgs/main/linux-64 ━━━━━━━━━╸━━━━━━━━━━━━━━━   0.0 B /  ??.?M

Next, we'll do is import two key data analysis modules: *pandas* and *numpy*.


In [2]:
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library

Let's download and import our primary Canadian Immigration dataset using *pandas*'s `read_excel()` method.


In [3]:
df_can = pd.read_excel(
    'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx',
    sheet_name='Canada by Citizenship',
    skiprows=range(20),
    skipfooter=2)

print('Data read into a pandas dataframe!')

Data read into a pandas dataframe!


Let's view the top 5 rows of the dataset using the `head()` function.


In [4]:
df_can.head()
# tip: You can specify the number of rows you'd like to see as follows: df_can.head(10) 

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  ...  2004  2005  2006  \
0    Southern Asia  902  Developing regions    16  ...  2978  3436  3009   
1  Southern Europe  901   Developed regions     1  ...  1450  1223   856   
2  Northern Africa  902  Developing regions    80  ...  3616  3626  4807   
3        Polynesia  902  Developing regions     0  ...     0     0     1   
4  Southern Europe  901   Developed regions     0  ...     0     0     1   

   2007  2008  2009  2010  2011  2012  2013  
0  2652  2111  1746  1758  2203  2635  2004  
1   702   560   716   561   539   620   603  
2  3623  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0     0  
4     1     0     0     0     0     1     1  

[5 rows x 43 columns]

We can also view the bottom 5 rows of the dataset using the `tail()` function.


In [5]:
df_can.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

When analyzing a dataset, it's always a good idea to start by getting basic information about your dataframe. We can do this by using the `info()` method.

This method can be used to get a short summary of the dataframe.


In [6]:
df_can.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


To get the list of column headers we can call upon the data frame's `columns` instance variable.


In [7]:
df_can.columns

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')

Similarly, to get the list of indices we use the `.index` instance variables.


In [8]:
df_can.index

RangeIndex(start=0, stop=195, step=1)

Note: The default type of intance variables `index` and `columns` are **NOT** `list`.


In [9]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


To get the index and columns as lists, we can use the `tolist()` method.


In [10]:
df_can.columns.tolist()

['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013]

In [ ]:
df_can.index.tolist()

In [11]:
print(type(df_can.columns.tolist()))
print(type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


To view the dimensions of the dataframe, we use the `shape` instance variable of it.


In [12]:
# size of dataframe (rows, columns)
df_can.shape

(195, 43)

**Note**: The main types stored in *pandas* objects are `float`, `int`, `bool`, `datetime64[ns]`, `datetime64[ns, tz]`, `timedelta[ns]`, `category`, and `object` (string). In addition, these dtypes have item sizes, e.g. `int64` and `int32`.


Let's clean the data set to remove a few unnecessary columns. We can use *pandas* `drop()` method as follows:


In [ ]:
# in pandas axis=0 represents rows (default) and axis=1 represents columns.
df_can.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
df_can.head(2)

Let's rename the columns so that they make sense. We can use `rename()` method by passing in a dictionary of old and new names as follows:


In [13]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.columns

Index([     'Type',  'Coverage',   'Country',      'AREA', 'Continent',
             'REG',    'Region',       'DEV',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013],
      dtype='object')

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:


In [14]:
df_can['Total'] = df_can.sum(axis=1)
df_can['Total']

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


0      65977
1      18433
2      72156
3       2774
4       2749
       ...  
190    99903
191     2719
192     5744
193     4392
194    11313
Name: Total, Length: 195, dtype: int64

We can check to see how many null objects we have in the dataset as follows:


In [15]:
df_can.isnull().sum()

Type         0
Coverage     0
Country      0
AREA         0
Continent    0
REG          0
Region       0
DEV          0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

Finally, let's view a quick summary of each column in our dataframe using the `describe()` method.


In [16]:
df_can.describe()

AREA          REG         DEV          1980          1981  \
count  195.000000   195.000000  195.000000    195.000000    195.000000   
mean   912.764103  1249.015385  901.753846    508.394872    566.989744   
std     13.082835  1185.526885    0.431878   1949.588546   2152.643752   
min    903.000000   905.000000  901.000000      0.000000      0.000000   
25%    903.000000   914.000000  902.000000      0.000000      0.000000   
50%    908.000000   922.000000  902.000000     13.000000     10.000000   
75%    922.000000   925.500000  902.000000    251.500000    295.500000   
max    935.000000  5501.000000  902.000000  22045.000000  24796.000000   

               1982          1983          1984         1985         1986  \
count    195.000000    195.000000    195.000000   195.000000   195.000000   
mean     534.723077    387.435897    376.497436   358.861538   441.271795   
std     1866.997511   1204.333597   1198.246371  1079.309600  1225.576630   
min        0.000000      0.000000      0.000000     0.000000     0.000000   
25%        0.000000      0.000000      0.000000     0.000000     0.500000   
50%       11.000000     12.000000     13.000000    17.000000    18.000000   
75%      275.000000    173.000000    181.000000   197.000000   254.000000   
max    20620.000000  10015.000000  10170.000000  9564.000000  9470.000000   

       ...          2005          2006          2007          2008  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...   1320.292308   1266.958974   1191.820513   1246.394872   
std    ...   4425.957828   3926.717747   3443.542409   3694.573544   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     28.500000     25.000000     31.000000     31.000000   
50%    ...    210.000000    218.000000    198.000000    205.000000   
75%    ...    832.000000    842.000000    899.000000    934.500000   
max    ...  42584.000000  33848.000000  28742.000000  30037.000000   

               2009          2010          2011          2012          2013  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1275.733333   1420.287179   1262.533333   1313.958974   1320.702564   
std     3829.630424   4462.946328   4030.084313   4247.555161   4237.951988   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       36.000000     40.500000     37.500000     42.500000     45.000000   
50%      214.000000    211.000000    179.000000    233.000000    213.000000   
75%      888.000000    932.000000    772.000000    783.000000    796.000000   
max    29622.000000  38617.000000  36765.000000  34315.000000  34129.000000   

               Total  
count     195.000000  
mean    35930.984615  
std     92040.242971  
min      2719.000000  
25%      3718.500000  
50%      8695.000000  
75%     24973.500000  
max    699242.000000  

[8 rows x 38 columns]

---
## *pandas* Intermediate: Indexing and Selection (slicing)<a id="3"></a>


### Select Column
**There are two ways to filter on a column name:**

Method 1: Quick and easy, but only works if the column name does NOT have spaces or special characters.
```python
    df.column_name               # returns series
```

Method 2: More robust, and can filter on multiple columns.

```python
    df['column']                  # returns series
```

```python 
    df[['column 1', 'column 2']]  # returns dataframe
```
---


Example: Let's try filtering on the list of countries ('Country').


In [17]:
df_can.Country  # returns a series

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Let's try filtering on the list of countries ('Country') and the data for years: 1980 - 1985.


In [18]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # returns a dataframe
# notice that 'Country' is string, and the years are integers. 
# for the sake of consistency, we will convert all column names to string later on.

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Select Row

There are main 2 ways to select rows:

```python
    df.loc[label]    # filters by the labels of the index/column
    df.iloc[index]   # filters by the positions of the index/column
```


Before we proceed, notice that the default index of the dataset is a numeric range from 0 to 194. This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corresponding index value.

This can be fixed very easily by setting the 'Country' column as the index using `set_index()` method.


In [19]:
df_can.set_index('Country', inplace=True)
# tip: The opposite of set is reset. So to reset the index, we can use df_can.reset_index()

In [20]:
df_can.head(3)

Type    Coverage  AREA Continent   REG           Region  \
Country                                                                      
Afghanistan  Immigrants  Foreigners   935      Asia  5501    Southern Asia   
Albania      Immigrants  Foreigners   908    Europe   925  Southern Europe   
Algeria      Immigrants  Foreigners   903    Africa   912  Northern Africa   

             DEV             DevName  1980  1981  ...  2005  2006  2007  2008  \
Country                                           ...                           
Afghanistan  902  Developing regions    16    39  ...  3436  3009  2652  2111   
Albania      901   Developed regions     1     0  ...  1223   856   702   560   
Algeria      902  Developing regions    80    67  ...  3626  4807  3623  4005   

             2009  2010  2011  2012  2013  Total  
Country                                           
Afghanistan  1746  1758  2203  2635  2004  65977  
Albania       716   561   539   620   603  18433  
Algeria      5393  4752  4325  3774  4331  72156  

[3 rows x 43 columns]

In [21]:
# optional: to remove the name of the index
df_can.index.name = None

Example: Let's view the number of immigrants from Japan (row 87) for the following scenarios:
    1. The full row data (all columns)
    2. For year 2013
    3. For years 1980 to 1985


In [ ]:
# 1. the full row data (all columns)
df_can.loc['Japan']

In [22]:
# alternate methods
df_can.iloc[87]

Type                Immigrants
Coverage            Foreigners
AREA                       935
Continent                 Asia
REG                        906
Region            Eastern Asia
DEV                        901
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004    

In [23]:
df_can[df_can.index == 'Japan']

Type    Coverage  AREA Continent  REG        Region  DEV  \
Japan  Immigrants  Foreigners   935      Asia  906  Eastern Asia  901   

                 DevName  1980  1981  ...  2005  2006  2007  2008  2009  2010  \
Japan  Developed regions   701   756  ...  1067  1212  1250  1284  1194  1168   

       2011  2012  2013  Total  
Japan  1265  1214   982  30449  

[1 rows x 43 columns]

In [24]:
# 2. for year 2013
df_can.loc['Japan', 2013]

982

In [25]:
# alternate method
# year 2013 is the last column, with a positional index of 36
df_can.iloc[87, 36]

1284

In [26]:
# 3. for years 1980 to 1985
df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]]

1980    701
1981    756
1982    598
1983    309
1984    246
1984    246
Name: Japan, dtype: object

In [27]:
# Alternative Method
df_can.iloc[87, [3, 4, 5, 6, 7, 8]]

Continent                 Asia
REG                        906
Region            Eastern Asia
DEV                        901
DevName      Developed regions
1980                       701
Name: Japan, dtype: object

**Exercise:** Let's view the number of immigrants from **Haiti** for the following scenarios: <br>1. The full row data (all columns) <br>2. For year 2000 <br>3. For years 1990 to 1995


In [28]:
df_can.loc['Haiti']
df_can.loc['Haiti', 2000]
df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]

1990    2379
1991    2829
1992    2399
1993    3655
1994    2100
1995    2014
Name: Haiti, dtype: object

<details><summary>Click here for a sample python solution</summary>

```python
   # 1. the full row data (all columns)
    df_can.loc['Haiti']
    #or
    df_can[df_can.index == 'Haiti']

    # 2. for year 2000
    df_can.loc['Haiti', 2000]
           
    #  3. for years 1990 to 1995
    df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]
 ```

</details>


Column names that are integers (such as the years) might introduce some confusion. For example, when we are referencing the year 2013, one might confuse that when the 2013th positional index. 

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.


In [32]:
df_can.columns = list(map(str, df_can.columns))
# [print (type(x)) for x in df_can.columns.values] #<-- uncomment to check type of column headers

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:


In [33]:
# useful for plotting later on
years = list(map(str, range(1980, 2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

**Exercise:** Create a list named 'year' using map function for years ranging from 1990 to 2013. <br>Then extract the data series from the dataframe df_can for Haiti using year list. 


In [37]:
  year = list(map(str, range(1990, 2014)))
  haiti = df_can.loc['Haiti', year] # passing in years 1990 - 2013
  haiti

1990    2379
1991    2829
1992    2399
1993    3655
1994    2100
1995    2014
1996    1955
1997    1645
1998    1295
1999    1439
2000    1631
2001    2433
2002    2174
2003    1930
2004    1652
2005    1682
2006    1619
2007    1598
2008    2491
2009    2080
2010    4744
2011    6503
2012    5868
2013    4152
Name: Haiti, dtype: object

<details><summary>Click here for a sample python solution</summary>

```python
    #The correct answer is:
    year = list(map(str, range(1990, 2014)))
    haiti = df_can.loc['Haiti', year] # passing in years 1990 - 2013
    
```
</details>


### Filtering based on a criteria <a id="4"></a>
To filter the dataframe based on a condition, we simply pass the condition as a boolean vector. 

For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).


In [38]:
# 1. create the condition boolean series
condition = df_can['Continent'] == 'Asia'
print(condition)

Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Viet Nam           True
Western Sahara    False
Yemen              True
Zambia            False
Zimbabwe          False
Name: Continent, Length: 195, dtype: bool


In [40]:
# 2. pass this condition into the dataFrame
df_can[condition]

Type    Coverage  AREA  \
Afghanistan                                     Immigrants  Foreigners   935   
Armenia                                         Immigrants  Foreigners   935   
Azerbaijan                                      Immigrants  Foreigners   935   
Bahrain                                         Immigrants  Foreigners   935   
Bangladesh                                      Immigrants  Foreigners   935   
Bhutan                                          Immigrants  Foreigners   935   
Brunei Darussalam                               Immigrants  Foreigners   935   
Cambodia                                        Immigrants  Foreigners   935   
China                                           Immigrants  Foreigners   935   
China, Hong Kong Special Administrative Region  Immigrants  Foreigners   935   
China, Macao Special Administrative Region      Immigrants  Foreigners   935   
Cyprus                                          Immigrants  Foreigners   935   
Democratic People's Republic of Korea           Immigrants  Foreigners   935   
Georgia                                         Immigrants  Foreigners   935   
India                                           Immigrants  Foreigners   935   
Indonesia                                       Immigrants  Foreigners   935   
Iran (Islamic Republic of)                      Immigrants  Foreigners   935   
Iraq                                            Immigrants  Foreigners   935   
Israel                                          Immigrants  Foreigners   935   
Japan                                           Immigrants  Foreigners   935   
Jordan                                          Immigrants  Foreigners   935   
Kazakhstan                                      Immigrants  Foreigners   935   
Kuwait                                          Immigrants  Foreigners   935   
Kyrgyzstan                                      Immigrants  Foreigners   935   
Lao People's Democratic Republic                Immigrants  Foreigners   935   
Lebanon                                         Immigrants  Foreigners   935   
Malaysia                                        Immigrants  Foreigners   935   
Maldives                                        Immigrants  Foreigners   935   
Mongolia                                        Immigrants  Foreigners   935   
Myanmar                                         Immigrants  Foreigners   935   
Nepal                                           Immigrants  Foreigners   935   
Oman                                            Immigrants  Foreigners   935   
Pakistan                                        Immigrants  Foreigners   935   
Philippines                                     Immigrants  Foreigners   935   
Qatar                                           Immigrants  Foreigners   935   
Republic of Korea                               Immigrants  Foreigners   935   
Saudi Arabia                                    Immigrants  Foreigners   935   
Singapore                                       Immigrants  Foreigners   935   
Sri Lanka                                       Immigrants  Foreigners   935   
State of Palestine                              Immigrants  Foreigners   935   
Syrian Arab Republic                            Immigrants  Foreigners   935   
Tajikistan                                      Immigrants  Foreigners   935   
Thailand                                        Immigrants  Foreigners   935   
Turkey                                          Immigrants  Foreigners   935   
Turkmenistan                                    Immigrants  Foreigners   935   
United Arab Emirates                            Immigrants  Foreigners   935   
Uzbekistan                                      Immigrants  Foreigners   935   
Viet Nam                                        Immigrants  Foreigners   935   
Yemen                                           Immigrants  Foreigners   935   

                                               Contin

In [41]:
# we can pass multiple criteria in the same line.
# let's filter for AreaNAme = Asia and RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

Type    Coverage  AREA Continent   REG  \
Afghanistan                 Immigrants  Foreigners   935      Asia  5501   
Bangladesh                  Immigrants  Foreigners   935      Asia  5501   
Bhutan                      Immigrants  Foreigners   935      Asia  5501   
India                       Immigrants  Foreigners   935      Asia  5501   
Iran (Islamic Republic of)  Immigrants  Foreigners   935      Asia  5501   
Maldives                    Immigrants  Foreigners   935      Asia  5501   
Nepal                       Immigrants  Foreigners   935      Asia  5501   
Pakistan                    Immigrants  Foreigners   935      Asia  5501   
Sri Lanka                   Immigrants  Foreigners   935      Asia  5501   

                                   Region  DEV             DevName  1980  \
Afghanistan                 Southern Asia  902  Developing regions    16   
Bangladesh                  Southern Asia  902  Developing regions    83   
Bhutan                      Southern Asia  902  Developing regions     0   
India                       Southern Asia  902  Developing regions  8880   
Iran (Islamic Republic of)  Southern Asia  902  Developing regions  1172   
Maldives                    Southern Asia  902  Developing regions     0   
Nepal                       Southern Asia  902  Developing regions     1   
Pakistan                    Southern Asia  902  Developing regions   978   
Sri Lanka                   Southern Asia  902  Developing regions   185   

                            1981  ...   2005   2006   2007   2008   2009  \
Afghanistan                   39  ...   3436   3009   2652   2111   1746   
Bangladesh                    84  ...   4171   4014   2897   2939   2104   
Bhutan                         0  ...      5     10      7     36    865   
India                       8670  ...  36210  33848  28742  28261  29456   
Iran (Islamic Republic of)  1429  ...   5837   7480   6974   6475   6580   
Maldives                       0  ...      0      0      2      1      7   
Nepal                          1  ...    607    540    511    581    561   
Pakistan                     972  ...  14314  13127  10124   8994   7217   
Sri Lanka                    371  ...   4930   4714   4123   4756   4547   

                             2010   2011   2012   2013   Total  
Afghanistan                  1758   2203   2635   2004   65977  
Bangladesh                   4721   2694   2640   3789   72906  
Bhutan                       1464   1879   1075    487   13214  
India                       34235  27509  30933  33087  699242  
Iran (Islamic Republic of)   7477   7479   7534  11291  183261  
Maldives                        4      3      1      1    7368  
Nepal                        1392   1129   1185   1308   17560  
Pakistan                     6811   7468  11227  12603  248938  
Sri Lanka                    4422   3309   3338   2394  155696  

[9 rows x 43 columns]

**Exercise:** Fetch the data where AreaName is 'Africa' and RegName is 'Southern Africa'. <br>Display the dataframe and find out how many instances are there?


In [42]:
 df_can[(df_can['Continent']=='Africa') & (df_can['Region']=='Southern Africa')]

Type    Coverage  AREA Continent  REG           Region  \
Botswana      Immigrants  Foreigners   903    Africa  913  Southern Africa   
Lesotho       Immigrants  Foreigners   903    Africa  913  Southern Africa   
Namibia       Immigrants  Foreigners   903    Africa  913  Southern Africa   
South Africa  Immigrants  Foreigners   903    Africa  913  Southern Africa   
Swaziland     Immigrants  Foreigners   903    Africa  913  Southern Africa   

              DEV             DevName  1980  1981  ...  2005  2006  2007  \
Botswana      902  Developing regions    10     1  ...     7    11     8   
Lesotho       902  Developing regions     1     1  ...     4     0     4   
Namibia       902  Developing regions     0     5  ...     6    19    13   
South Africa  902  Developing regions  1026  1118  ...   988  1111  1200   
Swaziland     902  Developing regions     4     1  ...     7     7     5   

              2008  2009  2010  2011  2012  2013  Total  
Botswana        28    15    42    53    64    76   3114  
Lesotho          1     8     7     1     0     6   2825  
Namibia         26    14    16    23    24    83   3038  
South Africa  1123  1188  1238   959  1243  1240  43286  
Swaziland        6    10     3    13    17    39   2906  

[5 rows x 43 columns]

<details><summary>Click here for a sample python solution</summary>

```python
    df_can[(df_can['Continent']=='Africa') & (df_can['Region']=='Southern Africa')]
```
</details>


### Sorting Values of a Dataframe or Series <a id="5"></a><br>
You can use the `sort_values()` function is used to sort a DataFrame or a Series based on one or more columns. <br>You to specify the column(s) by which you want to sort and the order (ascending or descending). Below is the syntax to use it:-<br><br>
```df.sort_values(col_name, axis=0, ascending=True, inplace=False, ignore_index=False)```<br><br>
col_nam - the column(s) to sort by. <br>
axis - axis along which to sort. 0 for sorting by rows (default) and 1 for sorting by columns.<br>
ascending - to sort in ascending order (True, default) or descending order (False).<br>
inplace - to perform the sorting operation in-place (True) or return a sorted copy (False, default).<br>
ignore_index - to reset the index after sorting (True) or keep the original index values (False, default).<br>


Let's sort out dataframe df_can on 'Total' column, in descending order to find out the top 5 countries that contributed the most to immigration to Canada. 


In [43]:
df_can.sort_values(by='Total', ascending=False, axis=0, inplace=True)
top_5 = df_can.head(5)
top_5

Type    Coverage  \
India                                               Immigrants  Foreigners   
China                                               Immigrants  Foreigners   
United Kingdom of Great Britain and Northern Ir...  Immigrants  Foreigners   
Philippines                                         Immigrants  Foreigners   
Pakistan                                            Immigrants  Foreigners   

                                                    AREA Continent   REG  \
India                                                935      Asia  5501   
China                                                935      Asia   906   
United Kingdom of Great Britain and Northern Ir...   908    Europe   924   
Philippines                                          935      Asia   920   
Pakistan                                             935      Asia  5501   

                                                                Region  DEV  \
India                                                    Southern Asia  902   
China                                                     Eastern Asia  902   
United Kingdom of Great Britain and Northern Ir...     Northern Europe  901   
Philippines                                         South-Eastern Asia  902   
Pakistan                                                 Southern Asia  902   

                                                               DevName   1980  \
India                                               Developing regions   8880   
China                                               Developing regions   5123   
United Kingdom of Great Britain and Northern Ir...   Developed regions  22045   
Philippines                                         Developing regions   6051   
Pakistan                                            Developing regions    978   

                                                     1981  ...   2005   2006  \
India                                                8670  ...  36210  33848   
China                                                6682  ...  42584  33518   
United Kingdom of Great Britain and Northern Ir...  24796  ...   7258   7140   
Philippines                                          5921  ...  18139  18400   
Pakistan                                              972  ...  14314  13127   

                                                     2007   2008   2009  \
India                                               28742  28261  29456   
China                                               27642  30037  29622   
United Kingdom of Great Britain and Northern Ir...   8216   8979   8876   
Philippines                                         19837  24887  28573   
Pakistan                                            10124   8994   7217   

                                                     2010   2011   2012  \
India                                               34235  27509  30933   
China                                               30391  28502  33024   
United Kingdom of Great Britain and Northern Ir...   8724   6204   6195   
Philippines                                         38617  36765  34315   
Pakistan                                             6811   7468  11227   

                                                     2013   Total  
India                                               33087  699242  
China                                               34129  662705  
United Kingdom of Great Britain and Northern Ir...   5827  554233  
Philippines                                         29544  514148  
Pakistan                                            12603  248938  

[5 rows x 43 columns]

**Exercise:** Find out top 3 countries that contributes the most to immigration to Canda in the year 2010. <br> Display the country names with the immigrant count in this year


In [44]:
  df_can.sort_values(by='2010', ascending=False, axis=0, inplace=True)
  top3_2010 = df_can['2010'].head(3)
  top3_2010

Philippines    38617
India          34235
China          30391
Name: 2010, dtype: int64

<details><summary>Click here for a sample python solution</summary>

```python
    df_can.sort_values(by='2010', ascending=False, axis=0, inplace=True)
    top3_2010 = df_can['2010'].head(3)
    top3_2010
```
</details>


Congratulations! you have learned how to wrangle data with Pandas. You will be using alot of these commands to preprocess the data before its can be used for data visualization.


### Thank you for completing this lab!


## Author

<a href="https://www.linkedin.com/in/aklson/" target="_blank">Alex Aklson</a>


### Other Contributors
[Jay Rajasekharan](https://www.linkedin.com/in/jayrajasekharan),
[Ehsan M. Kermani](https://www.linkedin.com/in/ehsanmkermani),
[Slobodan Markovic](https://www.linkedin.com/in/slobodan-markovic),
[Weiqing Wang](https://www.linkedin.com/in/weiqing-wang-641640133/),
[Dr. Pooja](https://www.linkedin.com/in/p-b28802262/)


## Change Log


|  Date (YYYY-MM-DD) | Version | Changed By    |  Change Description                   |
|--------------------|---------|---------------|---------------------------------------|
| 2023-06-08         | 2.5     | Dr. Pooja         |  Separated from original lab        |
| 2021-05-29         | 2.4     | Weiqing Wang  |  Fixed typos and code smells.         |
| 2021-01-20         | 2.3     | Lakshmi Holla |  Changed TOC cell markdown            |
| 2020-11-20         | 2.2     | Lakshmi Holla |  Changed IBM box URL                  |
| 2020-11-03         | 2.1     | Lakshmi Holla |  Changed URL and info method          |
| 2020-08-27         | 2.0     | Lavanya       |  Moved Lab to course repo in GitLab   |




## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
